nexus/reconfigurator/blippy/src/blippy.rs

```diff
@@ -0,0 +1,398 @@
+    ///
+    /// Many common blueprint use cases are likely to fail in some way if
+    /// performed with a blueprint reporting a `Fatal` note:
+    ///
+    /// * Uploading the blueprint to Nexus
+    /// * Attempting to execute the blueprint
+    /// * Attempting to generate a new child blueprint
+    Fatal,
+}
+
+impl fmt::Display for Severity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Severity::Fatal => write!(f, "FATAL"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Component {
+    Sled(SledUuid),
+}
+
+impl fmt::Display for Component {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Component::Sled(id) => write!(f, "sled {id}"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Kind<'a> {
+    /// Two running zones have the same underlay IP address.
+    DuplicateUnderlayIp {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+    },
+    /// A sled has two zones that are not members of the same sled subnet.
+    SledWithMixedUnderlaySubnets {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+    },
+    /// Two sleds are using the same sled subnet.
+    ConflictingSledSubnets {
+        other_sled: SledUuid,
+        subnet: Ipv6Subnet<SLED_PREFIX>,
+    },
+    /// An internal DNS zone has an IP that is not one of the expected rack DNS
+    /// subnets.
+    InternalDnsZoneBadSubnet {
+        zone: &'a BlueprintZoneConfig,
+        rack_dns_subnets: BTreeSet<DnsSubnet>,
+    },
+    /// Two running zones have the same external IP address.
+    DuplicateExternalIp {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+        ip: IpAddr,
+    },
+    /// Two running zones' NICs have the same IP address.
+    DuplicateNicIp {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+        ip: IpAddr,
+    },
+    /// Two running zones' NICs have the same MAC address.
+    DuplicateNicMac {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+        mac: MacAddr,
+    },
+    /// Two zones with the same durable dataset kind are on the same zpool.
+    ZoneDurableDatasetCollision {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+        zpool: ZpoolName,
+    },
+    /// Two zones with the same filesystem dataset kind are on the same zpool.
+    ZpoolFilesystemDatasetCollision {
+        zone1: &'a BlueprintZoneConfig,
+        zone2: &'a BlueprintZoneConfig,
+        zpool: ZpoolName,
+    },
+    /// One zpool has two datasets of the same kind.
+    ZpoolWithDuplicateDatasetKinds {
+        dataset1: &'a BlueprintDatasetConfig,
+        dataset2: &'a BlueprintDatasetConfig,
+        zpool: ZpoolUuid,
+    },
```

Comment:
I'm not opposed to more specific error kinds, but `ZpoolWithDuplicateDatasetKinds` will cover all three of these cases, right?

Reply:
I don't think so:

* `Zone*DatasetCollision` come from scanning `blueprint_zones`, from which we don't have a `BlueprintDatasetConfig` needed to construct a `ZpoolWithDuplicateDatasetKinds`
* `ZpoolWithDuplicateDatasetKinds` comes from scanning `blueprint_datasets` (so the inner fields are `BlueprintDatasetConfig`s, but these don't _necessarily_ have a corresponding `BlueprintZoneConfig` anywhere)

We could squish all of these down to one if we dropped the zone/dataset specifics to something like `DatasetKind`? That seems like it's dropping a lot of information, though.

Reply:
I think you're right, and I debated collapsing these, but I decided to keep the separate cases for explicitness. It's okay if we get some "double notes", and I think it's easier to reason about each of the checks locally if they're checking their side of the condition (e.g., zones check that their corresponding dataset exists and datasets check that their corresponding zone exists, even though strictly speaking we could drop one of those and not miss anything).